### DISCUSSION TITLE
Play 3.0.1 build/compile order

### DISCUSSION THREAD

**Discussion by nilsweinander** [playframework/playframework#12361] [Q&A] [answered]
I have a  problem caused by the order in which the code is compiled. The `build.sbt` for the Play project includes a dependency on another project:

```
ThisBuild / scalaVersion := "2.13.12"
Test / scalaVersion := "2.13.12"

ThisBuild / version := "2.0.0-alpha"

val emSrcProp = System.getProperty("org.kulturnav.entitymachine.src-path")
val emSrcPath = if (emSrcProp == null) "../entitymachine-ng" else emSrcProp

lazy val entitymachine = RootProject(file(emSrcPath))

lazy val root = (project in file("."))
  .enablePlugins(PlayJava)
  .settings(
    name := """kulturnav-2""",
    libraryDependencies ++= Seq(
      guice,
      javaJdbc,
      javaWs,
      caffeine,
      "be.objectify" %% "deadbolt-java" % "2.8.1",
      "org.openjdk.nashorn" % "nashorn-core" % "15.4",
      "org.lucee" % "txtmark" % "0.16.0"
    )
  ).dependsOn(entitymachine)

```

Building the project causes a compilation error in a Twirl template:

```
[error] /home/nils/idea-workspace/kulturnav/kulturnav-2/app/views/entity.scala.html:22:23: too many arguments (found 2, expected 1) for method defining: (t: T)(handler: T => Any): Any
```
on this template code:

```
@defining(
	languageUtils.preferredEntry(entity.caption, userSession.lang, messages.asJava).orElse(TextEntry.missing(userSession.lang)),
	schema.baseEntityType(entity.entityTypeName())
) { case(caption, baseType) =>

```

where the second argument to defining

`schema.baseEntityType(entity.entityTypeName())
`

is a call to the project **entitymachine** in the dependency (returning a String value). If I replace this with a constant String, the bulid is successful. If I then put the original code back, the build is successful. So, it seems like the Twirl template is compiled before the dependency project is compiled, so that code is not yet available. Or is there something else? Either way, what can I do about it?

**Comment by mkurz** [comment]
Try replacing this line
```sbt
).dependsOn(entitymachine)
```
with
```sbt
).aggregate(entitymachine).dependsOn(entitymachine)
```

so the project you dependon will be compiled also when compiling your main project.

**Reply by nilsweinander** [reply]
I have tried that now, but unfortunately it doesn't make any difference. I get the same error. My knowledge of sbt is very limited I'm afraid, but I will try to read up on this.

**Reply by nilsweinander** [reply]
So I googled a bit and found this:

"Note: aggregation will run the aggregated tasks in parallel and with no defined ordering between them. "

which would indicate that I cannot expect the dependency project to be compiled first. On the other hand:

"A project may depend on code in another project. This is done by adding a dependsOn method call. For example, if core needed util on its classpath, you would define core as:

lazy val core = project.dependsOn(util)

Now code in core can use classes from util. This also creates an ordering between the projects when compiling them; util must be updated and compiled before core can be compiled. "

which would mean that in my case **entitymachine** will compile first. I don't understand what is going on with the Twirl templates here.

**Reply by nilsweinander** [reply, marked as answer]
Never mind, I was barking up the wrong tree all the time. The actual problem is that Twirl doesn't accept empty parentheses for method calls without parameters:

```
@defining(
	languageUtils.preferredEntry(entity.caption, userSession.lang, messages.asJava).orElse(TextEntry.missing(userSession.lang)),
	schema.baseEntityType(entity.entityTypeName())
) { case(caption, baseType) =>

```
doesn't work (immediately)

```
@defining(
	languageUtils.preferredEntry(entity.caption, userSession.lang, messages.asJava).orElse(TextEntry.missing(userSession.lang)),
	schema.baseEntityType(entity.entityTypeName)
) { case(caption, baseType) =>

```
does.
